projects/electron-ngrx/src/lib/window-communication.service.ts

```diff
@@ -1,59 +1,112 @@
+import {filter, map, pluck, share, take, withLatestFrom} from 'rxjs/operators';
 import {NavigationEnd, Router} from '@angular/router';
+import {ipcRenderer, remote} from 'electron';
+
+interface WithSender {
+  data: any;
+  senderId: number;
+  massageId: number;
+}
+
+export interface Massage<T> {
+  reply: (...arg: any) => void;
+  data: T;
+}
+
+export const enum chancels {
+  replay = 'replay',
+  route = 'route',
+  id = 'id',
+  parent = 'parent'
+}
 
 @Injectable()
 export class WindowCommunicationService {
-  ipcRenderer;
-  remote;
+  ipcRenderer: typeof ipcRenderer;
+  remote: typeof remote;
+  private readonly myId: number;
+  private reply: Observable<any>;
 
   constructor(private router: Router) {
     const electron = (window as any).require('electron');
     this.ipcRenderer = electron.ipcRenderer;
     this.remote = electron.remote;
+    this.myId = electron.remote.getCurrentWindow().id;
+    this.reply = fromEvent(this.ipcRenderer, chancels.replay).pipe(map(this.fromElectronMassage), share());
   }
 
-  listenToChild() {
-    return fromEvent(this.ipcRenderer, 'send-to-parent')
-      .pipe(map(this.stripData));
+  listenToChild<T>(): Observable<Massage<T>> {
+    return this.listenIpcRenderer<T>(chancels.parent);
   }
 
-  listenToId() {
-    return fromEvent(this.ipcRenderer, 'send-to-id')
-      .pipe(map(this.stripData));
+  listenToId<T>(): Observable<Massage<T>> {
+    return this.listenIpcRenderer<T>(chancels.id);
   }
 
-  listenToRoute() {
-    return fromEvent(this.remote.ipcMain, 'send-to-route').pipe(
-      map(this.stripData),
-      withLatestFrom(this.getWindowRoute(), (data, url) => ({data, url})),
-      filter(({url, data}) => url === data.url),
-      map(({data}) => data.payload)
+  listenToRoute<T>(): Observable<{ data: T }> {
+    return fromEvent(this.remote.ipcMain, chancels.route).pipe(
+      map<any, { route: string, data: T }>(this.fromElectronMassage),
+      withLatestFrom(this.getWindowRoute(), (massage, myRoute) => ({massage, myRoute})),
+      filter(({myRoute, massage}) => myRoute === massage.route),
+      map(({massage}) => ({data: massage.data})),
     );
   }
 
-  private stripData([event, data]) {
+
+  sendToRoute<T>(route: string, data: T): void {
+    this.ipcRenderer.send(chancels.route, {data, route});
+  }
+
+  sendToParent<T, R>(data: T): Observable<R> {
+    const massageId = +new Date();
+    this.remote.getCurrentWindow().getParentWindow().webContents.send(chancels.parent, {
+      data,
+      senderId: this.myId,
+      massageId
+    });
+    return this.getReplyByMassageId<R>(massageId);
+  }
+
+
+  sendToId<T, R>(id: number, data: T): Observable<R> {
+    const massageId = +new Date();
+    this.sendById(chancels.id, id, {data, senderId: this.myId, massageId});
+    return this.getReplyByMassageId<R>(massageId);
+  }
+
+  private fromElectronMassage<T>([event, data]: [Event, T]): T {
     return data;
   }
 
+  private withReply<T>(massage: WithSender): Massage<T> {
```

Comment:
I've seen you using the name replay instead of reply alot of times.
Fix all of them to reply